### PR TITLE
Handling env reload response from native placeholder for failure case.

### LINKED ISSUE #9606: Prevent placeholder attempts with older versions of the SDK

### PR DESCRIPTION
**Pull request by kshyju**
If customer payload for dotnet isolated app is using an older version of worker package (which does not have the support to handle environment reload request via native placeholder, currently the request will crash. [We recently updated the NativePlaceholder to handle this case](https://github.com/Azure/azure-functions-dotnet-worker/pull/1955) and return a response for the env reload request. That version of Native placeholder is included in this PR. We also updated the code to handle the new response. If the failure is caused  due to customer payload having an older version of package, then we will execute our fallback logic, which is to shutdown the placeholder channel and start a new channel where we will load the customer assembly directly using dotnet.exe.

resolves #9606 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: test/DotNetIsolated60/DotNetIsolated60.csproj [hunk 1]
```diff
@@ -6,13 +6,15 @@
...
+    <FunctionsEnableExecutorSourceGen>True</FunctionsEnableExecutorSourceGen>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.18.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.19.0" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.13" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore" Version="1.0.0-preview2" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Storage" Version="6.0.0" />
-    <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.12.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.15.1" />
```
Comment (kshyju) [inline test/DotNetIsolated60/DotNetIsolated60.csproj:17]: Updated the existing test project to use the new worker.sdk project so that the "Placeholder hit" tests passes.

File: test/DotNetIsolated60/Program.cs [hunk 2]
```diff
@@ -11,14 +11,12 @@
 if (useProxy)
 {
     hostBuilder
-        .ConfigureFunctionsWebApplication()
-        .ConfigureGeneratedFunctionMetadataProvider();
```
Comment (kshyju) [inline test/DotNetIsolated60/Program.cs:15]: Removed this explicit extension method call in favor of the msbuild property (which does the auto registration of this methods).

File: test/DotNetIsolated60/DotNetIsolated60.csproj [hunk 3]
```diff
@@ -6,13 +6,15 @@
     <ImplicitUsings>enable</ImplicitUsings>
     <Nullable>enable</Nullable>
     <FunctionsEnableWorkerIndexing>True</FunctionsEnableWorkerIndexing>
+    <FunctionsAutoRegisterGeneratedMetadataProvider>True</FunctionsAutoRegisterGeneratedMetadataProvider>
+    <FunctionsEnableExecutorSourceGen>True</FunctionsEnableExecutorSourceGen>
```
Comment (kshyju) [inline test/DotNetIsolated60/DotNetIsolated60.csproj:10]: Added these msbuild properties for auto registration and removed the explicit extension method calls from Program.cs

File: test/DotNetIsolatedUnsupportedWorker/DotNetIsolatedUnsupportedWorker.csproj [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <FunctionsEnableWorkerIndexing>True</FunctionsEnableWorkerIndexing>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.19.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.13" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore" Version="1.0.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Storage" Version="6.0.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.14.0" />
```
Comment (kshyju) [inline test/DotNetIsolatedUnsupportedWorker/DotNetIsolatedUnsupportedWorker.csproj:15]: This is a new test project I added which references an older version of Worker.Sdk package which does not support placeholder. The `DotNetIsolated_PlaceholderMiss_UnsupportedWorkerPackage` test uses the publish output of this project as the function app payload.

File: test/EmptyScriptRoot/host.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (kshyju) [inline test/EmptyScriptRoot/host.json:1]: Could not check in an empty folder. Our host code automatically generates this file if not exist. So I decided to check in this so that the empty folder structure exist and can be used for `DotNetIsolated_PlaceholderMiss_EmptyScriptRoot` test.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 6]
```diff
@@ -376,13 +376,55 @@ internal void FunctionEnvironmentReloadResponse(FunctionEnvironmentReloadRespons
...
+            // Currently we support non fatal environment reload errors for dotnet isolated worker only.
+            if (workerRuntime != RpcWorkerConstants.DotNetIsolatedLanguageWorkerName)
+            {
+                return true;
+            }
+
+            var rpcException = response.Result.Exception;
+
+            // Currently the below 2 errors are considered as non fatal.
+            if (rpcException.Message.StartsWith("Microsoft.Azure.Functions.Worker.EnvironmentReloadNotSupportedException"))
```
Comment (kshyju) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: The native host will include this specific exception if the app payload uses an older version of worker package which does not support placeholders. https://github.com/Azure/azure-functions-dotnet-worker/blob/a70ead9e497c990d6c29b920f7ed1c3044f6c2f7/host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs#L64-L67

Comment (fabiocav) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: We should have this exception message be something that isn't specific to .NET. 

Ideally, this would be a flow any worker (any stack) can use.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 7]
```diff
@@ -376,13 +376,55 @@ internal void FunctionEnvironmentReloadResponse(FunctionEnvironmentReloadRespons
...
+            {
+                _workerChannelLogger.LogDebug(new EventId(422, ScriptConstants.PlaceholderMissDueToBitnessEventName),
```
Comment (fabiocav) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: We shouldn't log this here for the same reason. 

If we want to log details, we should pass that information from the worker and have the host log what came from it.

**Review by fabiocav — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 7 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Working on validating an updated version...

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 8]
```diff
@@ -376,13 +376,55 @@ internal void FunctionEnvironmentReloadResponse(FunctionEnvironmentReloadRespons
 
             if (res.Result.IsFailure(out Exception reloadEnvironmentVariablesException))
             {
-                _workerChannelLogger.LogError(reloadEnvironmentVariablesException, "Failed to reload environment variables");
-                _reloadTask.SetException(reloadEnvironmentVariablesException);
+                if (IsEnvironmentReloadFailureFatalError(res))
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: What are the cases where it's not fatal? I assumed that if the worker says it failed, we assume it's fatal. 

No one else even uses this result (b/c it never worked), so we can define the scenario...

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 9]
```diff
@@ -566,7 +581,9 @@ public Task SendFunctionEnvironmentReloadRequest()
                 FunctionEnvironmentReloadRequest = request
             });
 
-            return _reloadTask.Task;
+            var environmentReloadRequestResult = _reloadTask.Task;
+
+            return environmentReloadRequestResult;
```
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: is this change needed?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 10]
```diff
@@ -374,12 +374,18 @@ internal void FunctionEnvironmentReloadResponse(FunctionEnvironmentReloadRespons
 
             ApplyCapabilities(res.Capabilities, res.CapabilitiesUpdateStrategy.ToGrpcCapabilitiesUpdateStrategy());
 
-            if (res.Result.IsFailure(out Exception reloadEnvironmentVariablesException))
+            if (res.Result.IsFailure(IsUserCodeExceptionCapabilityEnabled(), out var reloadEnvironmentVariablesException))
             {
-                _workerChannelLogger.LogError(reloadEnvironmentVariablesException, "Failed to reload environment variables");
-                _reloadTask.SetException(reloadEnvironmentVariablesException);
+                if (res.Result.Exception is not null && reloadEnvironmentVariablesException is not null)
```
Comment (kshyju) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:379]: Checking `res.Result.Exception` for null here because I found when `res.Result.Exception` is null, we still create an instance of the RpcException with empty property values. Did not want to update that in this PR. Perhaps we can fix that later.

null-exception

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 7 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Pushed an update which logs the exception sent from the worker as it is. We set the "IsUserCodeException" to true so that message goes to AI logs.

**Review by brettsam — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 11]
```diff
@@ -414,6 +420,15 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
             _workerInitTask.TrySetResult(true);
         }
 
+        private bool IsUserCodeExceptionCapabilityEnabled()
+        {
+            var enableUserCodeExceptionCapability = string.Equals(
```
Comment (kshyju) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:425]: The native placeholder worker will send this capability.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 9 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: not needed. Reverted this in the latest push,.

**Review by fabiocav — APPROVED** [review]

**Review by soninaren — APPROVED** [review]

**Comment by safihamid** [comment]
>         private bool UsePlaceholderChannel(IRpcWorkerChannel channel)

@kshyju remind me again, do we need to shutdown the channel immediately or can wait like 5 sec? trying to see if this will have negative cold start impact on existing apps once we switch to opt out only

---
Refers to: src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs:160 in 42e4df3. [](commit_id = 42e4df33fd85f582d3375000a0466fdc2bfadade, deletion_comment = False)

**Comment by kshyju** [comment]
> > ```
> >     private bool UsePlaceholderChannel(IRpcWorkerChannel channel)
> > ```
> 
> @kshyju remind me again, do we need to shutdown the channel immediately or can wait like 5 sec? trying to see if this will have negative cold start impact on existing apps once we switch to opt out only
> 
> Refers to: src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs:160 in [42e4df3](https://github.com/Azure/azure-functions-host/commit/42e4df33fd85f582d3375000a0466fdc2bfadade). [](commit_id = [42e4df3](https://github.com/Azure/azure-functions-host/commit/42e4df33fd85f582d3375000a0466fdc2bfadade), deletion_comment = False)

Opened https://github.com/Azure/azure-functions-host/issues/9615 to investigate that.
